include/openthread/ip4.h

```diff
@@ -0,0 +1,106 @@
+/*
+ *  Copyright (c) 2022, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * @brief
+ *  This file defines the OpenThread IPv4 API for NAT64.
+ */
+
+#ifndef OPENTHREAD_IP4_H_
+#define OPENTHREAD_IP4_H_
+
+#include <openthread/message.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @addtogroup api-ip6
+ *
+ * @brief
+ *   This module includes functions that control IPv6 communication.
+ *
+ * @{
+ *
+ */
+
+#define OT_IP4_ADDRESS_SIZE 4 ///< Size of an IPv4 address (bytes)
+
+/**
+ * @struct otIp4Address
+ *
+ * This structure represents an IPv4 address.
+ *
+ */
+OT_TOOL_PACKED_BEGIN
+struct otIp4Address
+{
+    union OT_TOOL_PACKED_FIELD
+    {
+        uint8_t  m8[OT_IP4_ADDRESS_SIZE]; ///< 8-bit fields
+        uint32_t m32;                     ///< 32-bit representation
+    } mFields;
+} OT_TOOL_PACKED_END;
+
+/**
+ * This structure represents an IPv4 address.
+ *
+ */
+typedef struct otIp4Address otIp4Address;
+
+/**
+ * @struct otIp4Cidr
+ *
+ * This structure represents an IPv4 CIDR block.
+ *
+ */
+OT_TOOL_PACKED_BEGIN
+struct otIp4Cidr
```

Comment:
That means we should rename ip4.h into nat64.h and include `otBorderRouterSetIpv4CidrForNat64 `?

I think the structure will be clearer by moving `otBorderRouterSetIpv4CidrForNat64` into a new header however, I think `border_router_nat64.h` seems to be a better name for that header since we also have some logics for nat64 for the client side (`otDnsClientResolveIp4Address`), a single `nat64.h` is a bit confusing to me.

Also, I think `core/net/ip4*` is where we should place our IPv4 header & definitions. Then I think the mapped public header is `ip4.h`.

Note: I also have a change in `ip6.h` for adding `otIp6NewMessageForNat64`.

BTW, I have another PR https://github.com/openthread/openthread/pull/7824 with IPv4 related changes only.

Reply:
My suggestion is to have all APIs that are related to "NAT64 on a border router" into a a public header of its own. 
- Makes it easier for user of this feature to see all definitions/APIs. 
- Indicates that IPv4 (and all related definitions) require NAT64 (emphasizes that Thread is natively IPv6).
- Makes it clear all these definitions and APIs are available under specific feature and under build config `OPENTHREAD_CONFIG_BORDER_ROUTING_NAT64_ENABLE`.

Some earlier background:
- I actually added [`ip4.h`](https://github.com/openthread/openthread/commit/5fbde93d38c33bd5af1c5047353a56596d7af56a#diff-9f0856bbf465617f365571c371553c9bf7b2bf8a17376183e10ff5d65ba57817) public header earlier from https://github.com/openthread/openthread/pull/6334 
- @jwhui made a good point on why we may not want to include it ([comments](https://github.com/openthread/openthread/pull/6334#issuecomment-805983783)).

Reply:
Makes sense, I have updated #7824

Reply:
Resolved in 7824